src/ImageSharp/Diagnostics/MemoryDiagnostics.cs

```diff
@@ -47,18 +49,42 @@ public static class MemoryDiagnostics
             }
         }
 
+        internal static InteralMemoryDiagnostics Current
+        {
+            get
+            {
+                if (localInstance != null && localInstance.Value != null)
+                {
+                    return localInstance.Value;
+                }
+
+                return Default;
+            }
+
+            set
+            {
+                if (localInstance == null)
+                {
+                    lock (SyncRoot)
+                    {
+                        localInstance ??= new AsyncLocal<InteralMemoryDiagnostics>();
+                    }
+                }
+
+                localInstance.Value = value;
+            }
+        }
+
         /// <summary>
         /// Gets a value indicating the total number of memory resource objects leaked to the finalizer.
         /// </summary>
-        public static int TotalUndisposedAllocationCount => totalUndisposedAllocationCount;
+        public static int TotalUndisposedAllocationCount => Current.TotalUndisposedAllocationCount;
```

Comment:
`TotalUndisposedAllocationCount` was intended to report global allocation count. If I get the docs of [`AsyncLocal<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=net-6.0) right, the PR changes the semantics, so it becomes specific to an asynchronous execution flow, and will reset outside of the flow. I don't think this public behavior change is beneficial for the users of `MemoryDiagnostics.TotalUndisposedAllocationCount `. We need a separate API for `MemoryAllocatorValidator`.

Reply:
It way its setup it will only report the local context when its been overwritten in tests (internal api)   by default the backing fields are backs by the `MemoryDiagnostics.Default` static instance which will be used in production usage outside of testing.

Reply:
I see now, I missed the purpose of the branching. Though probably still better to keep the internal API as minimalistic and non-intrusive as possible.